### DISCUSSION TITLE
Rendering via API

### DISCUSSION THREAD

**Discussion by majorsauce** [bootstrap-email/bootstrap-email#103] [Ideas]
Hi,

in short:

**What ?**
Build a Web-Based API for rendering HTML

**Why ?**
Allows everyone from everywhere to render the HTML to send mails

Wouldn´t it be nice to allow rendering your templates using an Web-Based API as a service ?
You´ve already got the Online Editor, I was thinking about something similar but with actual let´s say REST-Endpoints that you can send your raw HTML to and it will return a rendered version of it.

Probably sounds too easy but I thought since you´ve already got the online editor set up this might be a way to render mails.
Obviously might lead to huge server loads depending on the user count ect. but it would enable live rendering from everywhere without a setup.

Cheers
Btw, love what you´re doing ❤️

**Comment by stuyam** [comment]
I do like this idea, I have thought about it a lot. Ultimately you hit the problem right on the nose, server load would be a huge consideration, as well as uptime and reliability. It's not out of the question but not right now. It would be very cool though. I wonder if it could use something like aws lambda or something to compile and handle load. Would have to charge for an API like that though because it wouldn't be cheap for me to maintain.

I would suggest using something like the CLI implementation for now to prerender you views in a programatic way that doesn't require copying and pasting from the editor: https://v1.bootstrapemail.com/docs/setup#command-line

**Comment by XMLDiva** [comment]
I was also thinking about an API of some sort. I'm running a .NET app and will totally be looking into how I could have this as an API or microservice that I can use in the few emails we do on the fly.

PS I also love what you're doing. I've had to do emails for many years and this would make things so much easier.

**Reply by ComtelJeremy** [reply]
Just as a side note, in case it helps you... we are running this for our e-mail templates as a build step...

- Our templates are in a git repo.
- When pushed to main or develop, the build step runs the CLI.
- The build step then places them into a blob storage after compiling where they can later be accessed.
- We then use the compiled html and inject it with values to send e-mails.

This helps since in our case the e-mails don't change often, and as a build step we don't have to think about the process. I may post the YAML we came up with and the shell script we use in the build step to run the CLI, if some would find that helpful. Perhaps @stuyam might even put it as an entry in the docs.

**Reply by majorsauce** [reply]
I‘m interested. We do also precompile and then fill it with values. I even built a little rendering/templating engine that will generate a mail by the Object you pass to it using a predefined template. Unfortunatle we still manually render the tremplate using the online editor whenever we change anything in it which is an annoying process

**Reply by XMLDiva** [reply]
I'm also interested. Some of our emails are precompiled templates, but we have a pseudo-shopping-cart table that has to be generated on the fly and has nested tables/grids depending on users actions.

**Comment by taylorchasewhite** [comment]
Curious what the latest on this is!

**Comment by MaaxGr** [comment]
Hey @stuyam,
i recently discovered this project on github: https://github.com/moveyourdigital/bootstrap-email-api
@moveyourdigital created a ruby on rails web api on top of this project. The image is also Docker Hub and i integrated it successfully in one of my projects.
Maybe we can integrate it into this repo directly?

Kind Regards
Max
